webapp/src/Entity/Submission.php

```diff
@@ -44,89 +42,85 @@ class Submission extends BaseApiEntity implements ExternalRelationshipEntityInte
      * @Serializer\Type("string")
      */
-    protected $submitid;
+    protected int $submitid;
 
     /**
-     * @var string
      * @ORM\Column(type="string", name="externalid", length=255,
      *     options={"comment"="Specifies ID of submission if imported from external CCS, e.g. Kattis",
      *              "collation"="utf8mb4_bin"},
      *     nullable=true)
      * @Serializer\Groups({"Nonstrict"})
      * @Serializer\SerializedName("external_id")
      */
-    protected $externalid;
+    protected ?string $externalid;
 
     /**
-     * @var double
+     * @var double|string
      * @ORM\Column(type="decimal", precision=32, scale=9, name="submittime", options={"comment"="Time submitted",
      *                             "unsigned"=true}, nullable=false)
      * @Serializer\Exclude()
      */
     private $submittime;
 
     /**
-     * @var boolean
      * @ORM\Column(type="boolean", name="valid",
      *     options={"comment"="If false ignore this submission in all scoreboard calculations",
      *              "default"="1"},
      *     nullable=false)
      * @Serializer\Exclude()
      */
-    private $valid = true;
+    private bool $valid = true;
 
     /**
-     * @var array
      * @ORM\Column(type="json", name="expected_results", length=255,
      *     options={"comment"="JSON encoded list of expected results - used to validate jury submissions"},
      *     nullable=true)
      * @Serializer\Exclude()
      */
-    private $expected_results;
+    private ?array $expected_results;
 
     /**
-     * @var string
      * @ORM\Column(type="string", name="entry_point", length=255,
      *     options={"comment"="Optional entry point. Can be used e.g. for java main class."},
      *     nullable=true)
      * @Serializer\Expose(if="context.getAttribute('domjudge_service').checkrole('jury')")
      */
-    private $entry_point;
+    private ?string $entry_point;
 
     /**
      * @ORM\ManyToOne(targetEntity="Contest", inversedBy="submissions")
      * @ORM\JoinColumn(name="cid", referencedColumnName="cid", onDelete="CASCADE")
      * @Serializer\Exclude()
      */
-    private $contest;
+    private Contest $contest;
 
     /**
      * @ORM\ManyToOne(targetEntity="Language", inversedBy="submissions")
      * @ORM\JoinColumn(name="langid", referencedColumnName="langid", onDelete="CASCADE")
      * @Serializer\Exclude()
      */
-    private $language;
+    private Language $language;
 
     /**
      * @ORM\ManyToOne(targetEntity="Team", inversedBy="submissions")
      * @ORM\JoinColumn(name="teamid", referencedColumnName="teamid", onDelete="CASCADE")
      * @Serializer\Exclude()
      */
-    private $team;
+    private Team $team;
 
     /**
      * @ORM\ManyToOne(targetEntity="User", inversedBy="submissions")
      * @ORM\JoinColumn(name="userid", referencedColumnName="userid", onDelete="CASCADE")
      * @Serializer\Exclude()
      */
-    private $user;
+    private ?User $user;
```

Comment:
Which submissions have no users? Shadow mode?

Reply:
Old ones at least. This is a very recent feature. And shadow indeed